File: Two Pointers/Medium/LET_no986_Interval List Intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        res = []
        i = j = 0
        m, n = len(firstList), len(secondList)

        while i < m and j < n:
            s1, e1 = firstList[i]
            s2, e2 = secondList[j]
            if s1 <= e2 and s2 <= e1:
                res.append([max(s1, s2), min(e2, e1)])
            if e1 <= e2:
                i += 1
            else:
                j += 1
        
        return res

File: Two Pointers/Medium/LET_no986_Interval List Intersections.cpp
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& firstList, vector<vector<int>>& secondList) {
        vector<vector<int>> res;
        int n1 = firstList.size(), n2 = secondList.size(), i = 0, j = 0;
        while (i < n1 && j < n2) {
            auto& first = firstList[i];
            auto& second = secondList[j];
            if (first[0] <= second[1] && second[0] <= first[1])
                res.push_back({max(first[0], second[0]), min(first[1], second[1])});
            if (first[1] <= second[1])
                i++;
            else
                j++;
        }
        
        return res;
    }
};